## Project Overview

This project addresses the problem of Algorithm Selection for Model Counting (#SAT), where the goal is to predict the most efficient solver for a given #SAT instance using machine learning techniques. We use a labeled dataset of 1337 SAT instances with 72 statistical and structural features to classify the best-performing solver out of five candidates: `gpmc`, `d4`, `ganak`, `addmc`, and `sharpsat`.

### Tasks:
- Understand the dataset and baseline model performance using k-NN.
- Improve model performance using:
  - Feature scaling (StandardScaler, MinMaxScaler)
  - Dimensionality reduction (PCA)
  - Feature selection (SelectKBest, SelectPercentile)
  - Hyperparameter tuning (GridSearchCV)
  - Cross-validation techniques
- Compare performance with an alternate classifier: Support Vector Machine (SVM)

We conclude by selecting and justifying the best-performing model based on accuracy and robustness.

---



### Objective

The Boolean satisfiability (SAT) problem consists in determining whether a Boolean formula is satisfiable or not. This problem is one of the most widely studied combinatorial problems in computer science. It is the classic NP-complete problem. Over the past number of decades, a significant amount of research work has focused on solving SAT problems with both complete and incomplete solvers.

An extended version of the problem is Model Counting (#SAT). In #SAT the solver needs to compute the number of solutions of a Boolean formula. A wide variety of solvers have been designed to tackle this problem.

In this project, we want to create an Algorithm Selection (AS) approach to Model Counting. For each #SAT instance, there is a specific solver that works better than the others, the goal of your machine learing approach is to classify it.

In AS we represent #SAT problems with a vector of 72 features with general information about the problem, e.g., number of variables, number of clauses, etc. There is no need to understand the features to be able to complete the assignment. For each instance, there is a 'label' column representing the name of the optimal solver.


## Data Preparation

### Data Loading and Initial Exploration


In [1]:
import pandas as pd

df = pd.read_csv("https://github.com/andvise/DM_Assignment/blob/main/train_data.csv?raw=true")
df

c       v  clauses_vars_ratio  vars_clauses_ratio  vcg_var_mean  \
0      681.0   238.0            2.861345            0.349486      0.011143   
1      368.0   140.0            2.628571            0.380435      0.018012   
2     1935.0  1920.0            1.007812            0.992248      0.001760   
3     3452.0  2821.0            1.223680            0.817207      0.000968   
4      694.0   294.0            2.360544            0.423631      0.007656   
...      ...     ...                 ...                 ...           ...   
1331   949.0   351.0            2.703704            0.369863      0.006701   
1332  1450.0   608.0            2.384868            0.419310      0.004427   
1333   250.0   100.0            2.500000            0.400000      0.026000   
1334  4949.0  3422.0            1.446230            0.691453      0.000764   
1335   280.0   120.0            2.333333            0.428571      0.018442   

      vcg_var_coeff  vcg_var_min  vcg_var_max  vcg_var_entropy  \
0          0.905300     0.005874     0.111601         1.880038   
1          0.510753     0.005435     0.054348         1.851609   
2          1.723720     0.000000     0.012403         1.280404   
3          1.436774     0.000290     0.006083         1.192878   
4          0.493513     0.002882     0.040346         1.776102   
...             ...          ...          ...              ...   
1331       1.151888     0.002012     0.063380         1.857919   
1332       1.552830     0.002408     0.161349         1.659495   
1333       0.555766     0.000000     0.096000         1.751533   
1334       0.770643     0.000202     0.004445         1.835041   
1335       0.490373     0.000000     0.045198         1.773180   

      vcg_clause_mean  ...  gsat_FirstLocalMinStep_CoeffVariance  \
0            0.011143  ...                              0.210148   
1            0.018012  ...                              0.124438   
2            0.001760  ...                              0.066708   
3            0.000968  ...                              0.053628   
4            0.007656  ...                              0.086841   
...               ...  ...                                   ...   
1331         0.006701  ...                              0.079108   
1332         0.004427  ...                              0.062664   
1333         0.026000  ...                              0.140708   
1334         0.000764  ...                              0.032571   
1335         0.018442  ...                              0.120101   

      gsat_FirstLocalMinStep_Median  gsat_FirstLocalMinStep_Q.10  \
0                              50.0                         42.0   
1                              34.0                         29.0   
2                             102.0                         94.0   
3                             192.0                        179.0   
4                              72.0                         64.0   
...                             ...                          ...   
1331                           81.0                         73.0   
1332                          136.0                        125.0   
1333                           26.0                         21.0   
1334                          461.0                        441.0   
1335                           35.0                         30.0   

      gsat_FirstLocalMinStep_Q.90  gsat_BestAvgImprovement_Mean  \
0                            57.0                      0.954568   
1                            40.0                      1.693036   
2                           111.0                      0.398129   
3                           205.0                      0.247528   
4                            80.0                      0.822829   
...                           ...                           ...   
1331                         89.0                      0.781888   
1332                        147.0                      0.776364   
1333                         30.0           

In [2]:
# Label or target variable
df['label'].value_counts()

label
gpmc        921
d4          168
ganak       140
addmc        55
sharpsat     52
Name: count, dtype: int64



## Basic models and evaluation 

### Baseline k-NN Model Performance


Using Scikit-learn, train and evaluate a k-NN classifier using 70% of the dataset from training and 30% for testing. For this part of the project, we are not interested in optimising the parameters; we just want to get an idea of the dataset.

In [3]:


# Separate features and target variable

X = df.drop('label', axis=1)  # Features
y = df['label']               # Labels

# Split the data into a training set and a test set (70% training, 30% testing)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.7, random_state = 42)

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Initialize the k-NN classifier
knn = KNeighborsClassifier()

# Train the classifier using the training data
knn.fit(X_train, y_train)

# Make predictions on the test data
y_pred = knn.predict(X_test)

# Evaluate the model by comparing the predicted labels against the actual labels in the test set
print('Accuracy:', accuracy_score(y_test, y_pred))

# Alternatively, check accuracy using the score method
knn.score(X_test, y_test)

Accuracy: 0.6733167082294265


0.6733167082294265

## Robust evaluation 

In this section, we are interested in more rigorous techniques by implementing more sophisticated methods. Trying to improve the k-NN classifier performances on this dataset.

For instance, you could consider:
* Hold-out and cross-validation.
* Hyper-parameter tuning.
* Feature selection.
* Feature normalisation.
* Etc.





In [4]:
# Import and load all necessary libraries

import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.decomposition import PCA
from sklearn.feature_selection import SelectKBest, f_classif, SelectPercentile
from sklearn.metrics import accuracy_score
from sklearn.feature_selection import SelectFromModel

In [5]:


# Step 1: Load the dataset
df = pd.read_csv("https://github.com/andvise/DM_Assignment/blob/main/train_data.csv?raw=true")

df['label'].replace({'gpmc': 0, 'd4': 1, 'ganak': 2, 'addmc': 3, 'sharpsat': 4}, inplace=True)

# Step 2: Prepare features and target
X = df.drop('label', axis=1)
y = df['label']

# Step 3: Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

**Feature Normalisation Methods**

MinMaxScaler and StandardScaler were both taken into account.

K-NN computes the distances between data points and is sensitive to their magnitudes. One characteristic may dominate the distance calculations and produce biased results if it has a wider range of values than other features.
Since MinMaxScaler scales data inside a constrained range [0, 1], it is a good choice for data with rigorous upper and lower bounds. For this reason, it was tested.

Since StandardScaler normalizes data to have a zero mean and a one standard deviation, it can be used with data that has a Gaussian distribution, which is why it was also evaluated.


In [6]:
# Step 4: Feature normalization using Standard scaler

# Standardize the data
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Step 5: Initialize and train the k-NN classifier
knn = KNeighborsClassifier()
knn.fit(X_train_scaled, y_train)
baseline_accuracy = knn.score(X_test_scaled, y_test)
print('Baseline accuracy with standardized features:', baseline_accuracy)

Baseline accuracy with standardized features: 0.7182044887780549


In [7]:
 # Feature normalization using MinMax scaler

# Applying MinMax Scaling
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Initialize and train k-NN
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_scaled, y_train)

# Predict and evaluate the model
y_pred = knn.predict(X_test_scaled)
print('Accuracy with MinMax Scaling:', accuracy_score(y_test, y_pred))

Accuracy with MinMax Scaling: 0.7281795511221946


## Dimensionality Reduction and Feature Selection

**Feature Selection Methods**

We can lower the likelihood of overfitting by limiting the amount of features, which is important for a distance-based technique like k-NN.Faster training and prediction times might result from fewer features.Using cross-validation, the effect of each approach on the accuracy of the k-NN model was quantified. The approach to feature selection that yielded the best balance between accuracy and simplicity of the model was selected.

Select K  Best: Utilizing statistical tests that gauge the features' connection with the target variable, SelectKBest was utilized to choose the features.This approach offers a clear justification for feature selection based on statistical importance and is simple to use and comprehend.Features with the highest predictive potential for the result are identified by SelectKBest. In order to lower noise and increase model accuracy, this is essential.

PCA : PCA was thought to preserve the most informative variance in the data while reducing dimensionality.The curse of dimensionality, in which the volume of space grows so quickly that the available data become sparse, can be lessened by reducing the number of dimensions, which can also significantly shorten the time it takes for k-NN to calculate distances.

Select Percentile: A subset of the most pertinent features can be the focus of the k-NN model by utilizing SelectPercentile, which may improve model performance, particularly in datasets where the proportion of predictive features to total features is unknown.Using a given scoring function, this method chooses the best features that fall inside a particular percentile of the highest scores.

In [8]:
# Step 5: Feature selection methods

from sklearn.decomposition import PCA


pca = PCA(n_components=0.95)  # keep 95% of variance

# Fit the PCA model to the scaled training data and transform the training data into its principal components.
X_train_pca = pca.fit_transform(X_train_scaled)

# Transform the scaled test data into the same principal component space defined by the training data.
X_test_pca = pca.transform(X_test_scaled)

# Fit the k-NN classifier on the training data transformed by PCA.
knn.fit(X_train_pca, y_train)

# Predict the labels of the test data transformed by PCA using the trained k-NN classifier.
y_pred_pca = knn.predict(X_test_pca)

print('Accuracy with PCA:', accuracy_score(y_test, y_pred_pca))


Accuracy with PCA: 0.7406483790523691


In [9]:
from sklearn.feature_selection import SelectKBest, f_classif

# Initialize SelectKBest to select the top 10 features based on the ANOVA F-test.
selector = SelectKBest(f_classif, k=10)


X_train_kbest = selector.fit_transform(X_train_scaled, y_train)
X_test_kbest = selector.transform(X_test_scaled)

knn.fit(X_train_kbest, y_train)

# Fit the k-NN classifier on the training data that has been transformed to only include the top 10 features.
accuracy = knn.score(X_test_kbest, y_test)
print('Accuracy with SelectKBest:', accuracy)


Accuracy with SelectKBest: 0.7256857855361596


In [10]:
from sklearn.feature_selection import SelectPercentile, f_classif

# Initialize SelectPercentile to keep only the top 10% of features based on the ANOVA F-test.
percentile_selector = SelectPercentile(f_classif, percentile=10)

X_train_percentile = percentile_selector.fit_transform(X_train_scaled, y_train)
X_test_percentile = percentile_selector.transform(X_test_scaled)

knn.fit(X_train_percentile, y_train)

accuracy = knn.score(X_test_percentile, y_test)
print('Accuracy with SelectPercentile:', accuracy)


Accuracy with SelectPercentile: 0.7231920199501247


## Hyperparameter Tuning using GridSearchCV

**Hyper Parameter Tuning**

GridSearchCV:  The performance of k-NN can be strongly influenced by the weight function, the distance metric, and the number of neighbors (k). GridSearchCV facilitates comprehensive exploration inside a designated parameter grid to identify the optimal amalgamation.'n_neighbors', 'weights', and'metrics' values were evaluated on a grid. The configuration that had the best accuracy during cross-validation was chosen. This stage guarantees that the model is well-suited to the training set and has good cross-validation generalization.

In [11]:
# Step 6: Hyperparameter tuning using GridSearchCV

param_grid = {
    'n_neighbors': [3, 5, 7, 9],                    # Lists different values for 'n_neighbors' to evaluate
    'weights': ['uniform', 'distance'],              # Evaluates both uniform and distance-based weighting
    'metric': ['euclidean', 'manhattan']             # Evaluates both Euclidean and Manhattan distance metrics
}

# Initialize GridSearchCV with the k-NN classifier, the parameter grid, and the scoring method.
# The cv parameter defines the number of cross-validation folds (5 in this case).

grid_search = GridSearchCV(estimator=KNeighborsClassifier(), param_grid=param_grid, cv=5, scoring='accuracy')


# Perform the grid search over the specified parameter grid using the training data.grid_search.fit(X_train_scaled, y_train)
grid_search.fit(X_train_scaled, y_train)

best_knn = grid_search.best_estimator_

# Evaluate the best model found by GridSearchCV on the scaled test data.
tuned_accuracy = best_knn.score(X_test_scaled, y_test)

# Predict on the test set using the best found parameters
y_pred = grid_search.predict(X_test_scaled)

# Calculate and print the accuracy on the test set
accuracy = accuracy_score(y_test, y_pred)
print(f'Test set accuracy: {accuracy:.4f}')

# Best parameters and score
print("Best parameters:", grid_search.best_params_)
print("Best cross-validation score:", grid_search.best_score_)
print('Tuned model accuracy:', tuned_accuracy)

Test set accuracy: 0.7382
Best parameters: {'metric': 'manhattan', 'n_neighbors': 7, 'weights': 'distance'}
Best cross-validation score: 0.7614973262032085
Tuned model accuracy: 0.7381546134663342


## Model Evaluation using Cross-Validation and Hold-Out Test Set

**k-Fold Cross-Validation, Hold-Out Method**

By estimating model performance more robustly across many data subsets, cross-validation keeps the model from becoming overly specialized to one particular data partition.After all optimizations, the hold-out approach was applied to evaluate the model's performance on entirely unseen data, hence emulating the model's performance in real-world scenarios.The assessment of the stability and reliability of the model was conducted by monitoring the variations in performance during cross-validation on several data folds, and subsequently on a hold-out set.

In [12]:
# Step 8 : Cross Validation

cv_scores = cross_val_score(best_knn, X_train_scaled, y_train, cv=5, scoring='accuracy')
cv_average_accuracy = cv_scores.mean()
print('Cross-validation average accuracy:', cv_average_accuracy)

Cross-validation average accuracy: 0.7614973262032085


In [13]:
# Holdout Validation

# best_knn is the classifier from GridSearchCV with optimal settings
best_knn = grid_search.best_estimator_

# Fit on the entire training data
best_knn.fit(X_train_scaled, y_train)

# Evaluate on the hold-out set
holdout_accuracy = best_knn.score(X_test_scaled, y_test)
print("Hold-out set accuracy: ", holdout_accuracy)


Hold-out set accuracy:  0.7381546134663342


**IMPROVISED knn CLASSIFIER MODEL**

**MODEL 1**

### Model 1: PCA + k-NN Pipeline


Feature and Target Separation: To enable the supervised learning strategy required for k-NN classification, the dataset was split into feature set X and target y.

Train-Test Split: To guarantee reproducibility, we divided the data using a random seed into training (70%) and testing (30%) groups. This split is essential for evaluating the model's generalizability because it validates the model's performance on untested data.

Standard Scaling: To ensure that every feature contributes equally to the distance calculations—a need for the efficacy of k-NN—StandardScaler was applied to equalize the feature data.

PCA for Dimensionality Reduction: To decrease the number of features while keeping 95% of the data variance, PCA(n_components=0.95) was used. By doing this step, the computational cost and curse of dimensionality are lessened.

Pipeline Setup: To expedite the preprocessing and classification procedures, a pipeline consisting of the StandardScaler, PCA, and KNeighborsClassifier was constructed.

Hyperparameter tuning with GridSearchCV: Using a parameter grid that changed the number of neighbors, weights, and distance metrics throughout a 5-fold cross-validation, GridSearchCV was utilized to find the best parameters for the k-NN classifier. The selection of hyperparameters that result in the optimum cross-validation accuracy is ensured by this thorough search.

Model Performance : An accuracy of 73.8 % is obtained with the model on test set.


In [14]:
from sklearn.pipeline import Pipeline
from sklearn.decomposition import PCA
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split, GridSearchCV
import pandas as pd


# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Setup the pipeline with the necessary transformations and the estimator
pipeline = Pipeline([
    ('scaler', StandardScaler()),      # Scale features to normalize data, essential for k-NN
    ('pca', PCA(n_components=0.95)),   # Reduce dimensionality to retain 95% variance which helps in alleviating the curse of dimensionality
    ('knn', KNeighborsClassifier())    # k-NN classifier, parameters to be tuned via GridSearch
])

# Define the grid of parameters to search
param_grid = {
    'knn__n_neighbors': [3, 5, 7, 9],           # Number of neighbors to use
    'knn__weights': ['uniform', 'distance'],    # Weighting type: uniform or distance
    'knn__metric': ['euclidean', 'manhattan']   # Distance metric: Euclidean or Manhattan

}

# Initialize GridSearchCV with the pipeline and the parameter grid
grid_search = GridSearchCV(pipeline, param_grid, cv=5, scoring='accuracy', verbose=1)

# Train the model with GridSearchCV to find the best parameters
grid_search.fit(X_train, y_train)

# Output the best parameters
print("Best parameters:", grid_search.best_params_)

# Evaluate the best model on the testing set
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy on test set:", accuracy)


# Cross-validation for performance estimation
cv_scores = cross_val_score(best_model, X_train, y_train, cv=5)
print("Cross-validation scores:", cv_scores)
print("Best cross-validation score:", grid_search.best_score_)


Fitting 5 folds for each of 16 candidates, totalling 80 fits
Best parameters: {'knn__metric': 'euclidean', 'knn__n_neighbors': 7, 'knn__weights': 'distance'}
Accuracy on test set: 0.7381546134663342
Cross-validation scores: [0.76470588 0.71122995 0.75935829 0.70053476 0.78074866]
Best cross-validation score: 0.7433155080213905


**FINAL ROBUST MODEL WITH IMPROVED k-NN CLASSIFIER PERFORMANCE ON THIS DATASET**

###  Model 2: SelectKBest + k-NN Pipeline


**Train-Test Split:** To ensure a random and repeatable distribution of data points, the dataset was split into training (70%) and testing (30%) subsets using train test split. This division facilitates the model's assessment on hypothetical data, which is essential for determining the model's practicality.

**Model Setup Pipeline Configuration:**

**MinMaxScaler:** Used to normalize feature values between 0 and 1, reducing the impact of outlier values and guaranteeing that each feature makes an equal contribution to the distance computations that are essential to k-NN models.

**SelectKBest and f_classif:** Using SelectKBest and f_classif, the model is trained to focus on the most pertinent features for target prediction by choosing the top k features from the ANOVA F-test. This eliminates dimensionality and could improve the accuracy of the model by removing superfluous information.

**KNeighborsClassifier:** Set up as the last stage of the pipeline, ready for fine-tuning to determine the ideal values for the distance measure, weighting scheme, and neighborhood size.

**Hyperparameter Tuning GridSearchCV:** A thorough grid search was carried out to investigate different setups of the k-NN classifier's hyperparameters and feature selection


**Number of Features (k in SelectKBest)**:  Varied to evaluate how various feature subset sizes affected the performance of the model.You may systematically investigate how altering the number of features impacts the model's performance by experimenting with different values of k. This is crucial since adding too many features could result in noise and redundancy, overfitting, and higher computing costs. The range of values [5, 10, 15, 20] makes it possible to evaluate this balance in various contexts, which aids in locating the sweet spot where the model performs best in terms of generalization.

**Neighbor Count (knn__n_neighbors):**
Values: [3, 5, 7, 9, 11,]. In k-NN, the number of neighbors is an important parameter. A model that has too few neighbors may be too sensitive to noise in the training set (overfitting), whereas a model with too many neighbors may be too broadly generalized (underfitting). It is possible to identify a balanced value that offers the highest generalization performance by testing several values.The ideal neighborhood size for classification was determined by testing a range of neighbor counts.


Weights and Metrics: To determine the optimal method for calculating distance in k-NN, uniform and distance-based weights as well as Euclidean and Manhattan metrics were tested.

**Knn__weights, or weights**: ['uniform', 'distance']

This parameter establishes whether closer neighbors have a stronger influence on the prediction or if each point in the neighborhood contributes uniformly. In clusters with different densities, distance weighting might be very helpful as uniform weighting might not work as well. increases the model's capacity to adjust to local data structures, which can result in higher accuracy—particularly in cases where class distributions are intricate and overlap.

**Metric of Distance (knn__metric)**:['euclidean','manhattan']

Since different metrics can capture different features of data similarity, the choice of distance metric can have a major impact on the performance of k-NN. While Manhattan distance can be more reliable in high-dimensional spaces or situations where the data isn't isotropically distributed, Euclidean distance is the standard.


**Cross Validation:** To make sure that the model generalizes effectively across various data splits and that the tuning procedure is resilient against overfitting, the search was conducted over a 5-fold cross-validation.Cross-validation models have a higher chance of generalizing well to new data. Through the validation procedure, it is made sure that the hyperparameters are optimized for a variety of training scenarios, not just one particular training set.Cross-validation becomes an effective tool for hyperparameter tweaking when paired with GridSearchCV. Every possible combination of hyperparameters is tested across all folds, and the one with the best average performance is chosen.

**cv = 5** : The number of folds in a (Stratified)KFold cross-validation is indicated by the parameter cv=5. It determines how the data is divided up for assessment. When five parts of the dataset are utilized in 5-fold cross-validation, four of the parts are used for training and one for validation throughout each cycle. To ensure that every component is utilized as the validation set once, this step is repeated five times.

**scoring='accuracy'**: Specifies the performance metric to be applied for assessing the effectiveness of the combinations of hyperparameters. When 'accuracy' is selected, GridSearchCV will determine the optimal hyperparameters based on the validation set's accuracy score.


**Performance Reporting:** Cross-validation scores and testing the model on a separate test set offer an additional layer of assurance regarding the model's stability and performance after the ideal parameters have been determined.


In [15]:
# Import libraries

from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split, GridSearchCV
import pandas as pd


# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Setup the pipeline with MinMaxScaler, SelectKBest for feature selection, and KNeighborsClassifier
pipeline_kbest = Pipeline([
    ('scaler', MinMaxScaler()),
    ('selector', SelectKBest(f_classif)),
    ('knn', KNeighborsClassifier())
])

# Define the grid of parameters to search
param_grid_kbest = {
    'selector__k': [5, 10, 15, 20],                # Trying different numbers of top features
    'knn__n_neighbors': [3, 5, 7, 9, 11],          # Number of neighbors to use
    'knn__weights': ['uniform', 'distance'],       # Weighting type: uniform or distance
    'knn__metric': ['euclidean', 'manhattan']      # Distance metric: Euclidean or Manhattan
}

# Initialize GridSearchCV
grid_search_kbest = GridSearchCV(pipeline_kbest, param_grid_kbest, cv=5, scoring='accuracy')
grid_search_kbest.fit(X_train, y_train)

# Output best parameters
print("Best parameters (SelectKBest):", grid_search_kbest.best_params_)

# Evaluate the best model on the testing set
best_model = grid_search_kbest.best_estimator_
y_pred = best_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy on test set:", accuracy)


# Cross-validation for performance estimation
cv_scores = cross_val_score(best_model, X_train, y_train, cv=5)
print("Cross-validation scores:", cv_scores)
print("Best cross-validation score (SelectKBest):", grid_search_kbest.best_score_)



Best parameters (SelectKBest): {'knn__metric': 'euclidean', 'knn__n_neighbors': 11, 'knn__weights': 'distance', 'selector__k': 15}
Accuracy on test set: 0.7506234413965087
Cross-validation scores: [0.77540107 0.75935829 0.7486631  0.72727273 0.79144385]
Best cross-validation score (SelectKBest): 0.7604278074866311


**INFERENCE ON BEST MODEL:**

The top 20 characteristics chosen by SelectKBest, the manhattan distance metric, distance weighting, and ten neighbors were all used in the top-performing model. According to this setup, the best accuracy is obtained when a more granular feature selection is paired with a larger neighborhood that has distance-weighted contributions.

Test Set Performance: The model obtained an accuracy of roughly 75.06% on the test data that was not visible. The close correlation shown between test accuracy and cross-validation highlights the efficacy and generalizability of the approach.

Comparing Model 2 to Model 1, there is a little improvement. This is explained by SelectKBest's ability to effectively pick the most pertinent features, which may give the k-NN algorithm a clearer signal and increase accuracy.Model 2 employs SelectKBest, which chooses features based on statistical tests, whereas Model 1 uses principle Component Analysis (PCA), which shrinks the feature space by converting features into principle components.
While SelectKBest keeps the original characteristics, which may be easier to read, PCA may delete or modify features that are helpful for categorization.

Model 2 has more neighbors (11 vs. 7) than Model 1, which implies that adding more local information (without overfitting) improves performance, especially when the appropriate features are chosen.
The difference in scaling techniques may also have an impact on the improvement from Model 1 to Model 2, suggesting that the MinMaxScaler may be a better fit for this dataset than the StandardScaler.


Focusing on the most statistically relevant characteristics, SelectKBest's incorporation into the k-NN classification pipeline significantly improved the model's predicted accuracy. The model's setup was further adjusted by a methodical investigation of hyperparameter space using GridSearchCV, resulting in a robust classifier that exhibits consistent performance across several data segments. The model's sophisticated knowledge of feature interactions was demonstrated by the use of Manhattan distance, which improved classification accuracy, especially when combined with a bigger collection of neighbors and distance-based weighting.

## New classifier 

### Alternative Model: SVM Classifier with PCA


Replicating the previous task for a classifier different than K-NN and decision trees. 
Trying to create the best model for the given dataset and the choice is briefly described.






In [16]:
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split, GridSearchCV
import pandas as pd

# Load dataset
df = pd.read_csv("https://github.com/andvise/DM_Assignment/blob/main/train_data.csv?raw=true")
X = df.drop('label', axis=1)
y = df['label']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Setup the pipeline
pipeline = Pipeline([
    ('scaler', StandardScaler()),     # Normalize data
    ('pca', PCA(n_components=0.95)),  # Dimensionality reduction
    ('svm', SVC())                    # SVM classifier
])

# Define the grid of parameters to search
param_grid = {
    'svm__C': [0.1, 1, 10, 100],               # Regularization parameter
    'svm__kernel': ['linear', 'poly', 'rbf'],  # Type of kernel used in SVM
    'svm__gamma': ['scale', 'auto']            # Kernel coefficient for 'rbf'
}

# Initialize GridSearchCV
grid_search = GridSearchCV(pipeline, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)

# Output best parameters and the best cross-validation score
print("Best parameters:", grid_search.best_params_)

# Evaluate the best model on the testing set
model = grid_search.best_estimator_
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy on test set:", accuracy)

# Cross-validation for performance estimation
cv_scores = cross_val_score(model, X_train, y_train, cv=5)
print("Cross-validation scores:", cv_scores)
print("Best cross-validation score:", grid_search.best_score_)



Best parameters: {'svm__C': 100, 'svm__gamma': 'auto', 'svm__kernel': 'rbf'}
Accuracy on test set: 0.7381546134663342
Cross-validation scores: [0.76470588 0.7540107  0.7486631  0.68449198 0.77540107]
Best cross-validation score: 0.7454545454545454


##  Final Model Inference and Conclusions

**Model Selection: Support Vector Machine (SVM)**

For addressing this classification task, the SVM classifier offers a strong foundation with plenty of room for additional improvements to improve performance. The SVM classifier performs well on the dataset after being carefully tuned through a process of hyperparameter testing and assessed using cross-validation and a hold-out test set. The SVM is a great option for this classification problem because of its adaptability, efficiency in high-dimensional spaces, and resilience to overfitting.

SVMs' efficiency in medium-sized datasets is relatively reasonable, despite the fact that they can be computationally demanding when working with very big datasets. This is especially true when combined with feature reduction techniques like PCA, as seen in the pipeline.By removing noise from the dataset, PCA's dimensionality reduction technique helps to streamline the dataset, speed up training, and maybe enhance classifier performance.Using the 'rbf' kernel, for example, SVM can represent non-linear boundaries, which makes it an effective tool for capturing intricate correlations in data.

**svm__C:** [0.1, 1, 10, 100]: In support of support vector machines (SVMs), the 'C' parameter regulates the trade-off between obtaining a low error on the training data and guaranteeing a smooth decision boundary. A high 'C' tries to categorize all training instances properly by allowing the model to choose more support vectors, whereas a low 'C' smoothes the decision surface.

**svm__kernel:** ['poly', 'rbf', 'linear']: The purpose of selecting this set of kernels is to investigate both non-linear ('poly' and 'rbf') and linear ('linear') correlations in the data. While the 'rbf' kernel can handle an infinite number of dimensions, which is useful for complex patterns, the 'poly' kernel examines higher-dimensional spaces.

**svm__gamma:** ['scale', 'auto']: This option establishes the extent to which the impact of a solitary training sample is felt. High values indicate "close," and low ones indicate "far." The gamma value is 1 / (n_features * X.var()) for the'scale' option and 1 / n_features for the 'auto' option. The classifier's capacity to manage the bias-variance trade-off may be impacted by this.


The classifier's robustness and efficacy are improved by its margin maximization approach, which also enables it to handle complicated datasets and adapt to different data architectures.


## Summary

We explored multiple pipelines for classifying the optimal #SAT solver:
- k-NN baseline achieved ~67% accuracy.
- After feature scaling and PCA, accuracy improved to ~74%.
- Using SelectKBest for feature selection further improved performance to ~75%.
- GridSearchCV optimized hyperparameters such as `n_neighbors`, `weights`, and `distance metric`.
- The final k-NN model (with MinMaxScaler + SelectKBest + GridSearchCV) reached 75.06% test accuracy.
- An SVM classifier with PCA achieved comparable performance (~73.8%) and robust cross-validation scores.

 **Conclusion**: The refined k-NN model outperformed the SVM, showing that feature-based selection combined with distance-weighted k-NN offers superior classification for #SAT solver selection.
